[CMake] `find_package(LLVM CONFIG)` may fail when the project does not use C

1. Build LLVM 14 rc1 with `LLVM_ENABLE_FFI` and `LLVM_ENABLE_TERMINFO` options enabled.
2. Create a C++ only CMake project :
```cmake
cmake_minimum_required(VERSION 3.16)  # just to silence the warning when this line is missing
project(test LANGUAGES CXX)
find_package(LLVM CONFIG REQUIRED)
```
3. Configure this project:
```console
$ cmake -Bbuild . -DLLVM_DIR=/root/llvm-project-14.0.0rc1.src/llvm/build/lib/cmake/llvm/
-- The CXX compiler identification is GNU 12.0.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/share/cmake/Modules/Internal/CheckSourceCompiles.cmake:44 (message):
  check_source_compiles: C: needs to be enabled before use.
Call Stack (most recent call first):
  /usr/share/cmake/Modules/CheckCSourceCompiles.cmake:76 (cmake_check_source_compiles)
  /root/llvm-project-14.0.0rc1.src/llvm/build/lib/cmake/llvm/FindFFI.cmake:44 (check_c_source_compiles)
  /root/llvm-project-14.0.0rc1.src/llvm/build/lib/cmake/llvm/LLVMConfig.cmake:145 (find_package)
  CMakeLists.txt:3 (find_package)


-- Could NOT find FFI (missing: HAVE_FFI_CALL)
CMake Error at /usr/share/cmake/Modules/Internal/CheckSourceCompiles.cmake:44 (message):
  check_source_compiles: C: needs to be enabled before use.
Call Stack (most recent call first):
  /usr/share/cmake/Modules/CheckCSourceCompiles.cmake:76 (cmake_check_source_compiles)
  /root/llvm-project-14.0.0rc1.src/llvm/build/lib/cmake/llvm/FindTerminfo.cmake:21 (check_c_source_compiles)
  /root/llvm-project-14.0.0rc1.src/llvm/build/lib/cmake/llvm/LLVMConfig.cmake:152 (find_package)
  CMakeLists.txt:3 (find_package)


-- Could NOT find Terminfo (missing: Terminfo_LINKABLE)
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.11")
-- Found LibXml2: /usr/lib64/libxml2.so (found version "2.9.12")
-- Configuring incomplete, errors occurred!
See also "/root/test/build/CMakeFiles/CMakeOutput.log".
```

Changing the `project(test LANGUAGES CXX)` line to `project(test LANGUAGES C CXX)` makes the error go away, but that's just a work-around than a solution.

CMake version: 3.22.2
Introduced by: https://reviews.llvm.org/D114327

--- a/llvm/cmake/modules/FindFFI.cmake
+++ b/llvm/cmake/modules/FindFFI.cmake
@@ -40,7 +40,7 @@ if(FFI_LIBRARIES)
   include(CMakePushCheckState)
   cmake_push_check_state()
   list(APPEND CMAKE_REQUIRED_LIBRARIES ${FFI_LIBRARIES})
-  set(SRC [=[
+  set(HAVE_FFI_CALL_SRC [=[
     #ifdef __cplusplus
     extern "C" {
     #endif
@@ -54,10 +54,10 @@ if(FFI_LIBRARIES)
     ]=])
   if(DEFINED CMAKE_C_COMPILER)
     include(CheckCSourceCompiles)
-    check_c_source_compiles("${SRC}" HAVE_FFI_CALL)
+    check_c_source_compiles("${HAVE_FFI_CALL_SRC}" HAVE_FFI_CALL)
   else()
     include(CheckCXXSourceCompiles)
-    check_cxx_source_compiles("${SRC}" HAVE_FFI_CALL)
+    check_cxx_source_compiles("${HAVE_FFI_CALL_SRC}" HAVE_FFI_CALL)
   endif()
   cmake_pop_check_state()
 endif()

--- a/llvm/cmake/modules/FindFFI.cmake
+++ b/llvm/cmake/modules/FindFFI.cmake
@@ -38,15 +38,27 @@ find_library(FFI_LIBRARIES ffi PATHS ${FFI_LIBRARY_DIR})
 
 if(FFI_LIBRARIES)
   include(CMakePushCheckState)
-  include(CheckCSourceCompiles)
   cmake_push_check_state()
   list(APPEND CMAKE_REQUIRED_LIBRARIES ${FFI_LIBRARIES})
-  check_c_source_compiles("
+  set(SRC [=[
+    #ifdef __cplusplus
+    extern "C" {
+    #endif
     struct ffi_cif;
     typedef struct ffi_cif ffi_cif;
     void ffi_call(ffi_cif *cif, void (*fn)(void), void *rvalue, void **avalue);
-    int main(void) { ffi_call(0, 0, 0, 0); }"
-    HAVE_FFI_CALL)
+    #ifdef __cplusplus
+    }
+    #endif
+    int main(void) { ffi_call(0, 0, 0, 0); }
+    ]=])
+  if(DEFINED CMAKE_C_COMPILER)
+    include(CheckCSourceCompiles)
+    check_c_source_compiles("${SRC}" HAVE_FFI_CALL)
+  else()
+    include(CheckCXXSourceCompiles)
+    check_cxx_source_compiles("${SRC}" HAVE_FFI_CALL)
+  endif()
   cmake_pop_check_state()
 endif()
 

--- a/llvm/cmake/modules/FindTerminfo.cmake
+++ b/llvm/cmake/modules/FindTerminfo.cmake
@@ -17,7 +17,7 @@ if(Terminfo_LIBRARIES)
   include(CMakePushCheckState)
   cmake_push_check_state()
   list(APPEND CMAKE_REQUIRED_LIBRARIES ${Terminfo_LIBRARIES})
-  set(SRC [=[
+  set(Terminfo_LINKABLE_SRC [=[
     #ifdef __cplusplus
     extern "C" {
     #endif
@@ -29,10 +29,10 @@ if(Terminfo_LIBRARIES)
     ]=])
   if(DEFINED CMAKE_C_COMPILER)
     include(CheckCSourceCompiles)
-    check_c_source_compiles("${SRC}" Terminfo_LINKABLE)
+    check_c_source_compiles("${Terminfo_LINKABLE_SRC}" Terminfo_LINKABLE)
   else()
     include(CheckCXXSourceCompiles)
-    check_cxx_source_compiles("${SRC}" Terminfo_LINKABLE)
+    check_cxx_source_compiles("${Terminfo_LINKABLE_SRC}" Terminfo_LINKABLE)
   endif()
   cmake_pop_check_state()
 endif()

--- a/llvm/cmake/modules/FindTerminfo.cmake
+++ b/llvm/cmake/modules/FindTerminfo.cmake
@@ -15,13 +15,25 @@ find_library(Terminfo_LIBRARIES NAMES terminfo tinfo curses ncurses ncursesw)
 
 if(Terminfo_LIBRARIES)
   include(CMakePushCheckState)
-  include(CheckCSourceCompiles)
   cmake_push_check_state()
   list(APPEND CMAKE_REQUIRED_LIBRARIES ${Terminfo_LIBRARIES})
-  check_c_source_compiles("
+  set(SRC [=[
+    #ifdef __cplusplus
+    extern "C" {
+    #endif
     int setupterm(char *term, int filedes, int *errret);
-    int main(void) { return setupterm(0, 0, 0); }"
-    Terminfo_LINKABLE)
+    #ifdef __cplusplus
+    }
+    #endif
+    int main(void) { return setupterm(0, 0, 0); }
+    ]=])
+  if(DEFINED CMAKE_C_COMPILER)
+    include(CheckCSourceCompiles)
+    check_c_source_compiles("${SRC}" Terminfo_LINKABLE)
+  else()
+    include(CheckCXXSourceCompiles)
+    check_cxx_source_compiles("${SRC}" Terminfo_LINKABLE)
+  endif()
   cmake_pop_check_state()
 endif()
 

